frontend/scripts/react-components/dashboard-element/dashboard-widget/table-modal/table-modal.selectors.js

```diff
@@ -0,0 +1,149 @@
+import { format } from 'd3-format';
+import { CHART_TYPES } from 'constants';
+
+const getMeta = (state, { meta }) => meta || null;
+const getData = (state, { data }) => data || null;
+const getChartType = (state, { chartType }) => chartType || null;
+const returnNameArrayIfNotEmpty = value =>
+  isEmpty(value) ? null : Object.values(value).map(v => v.name);
+const getCountriesPanel = state =>
+  returnNameArrayIfNotEmpty(
+    state.dashboardElement && state.dashboardElement.countriesPanel.activeItems
+  );
+const getCommoditiesPanel = state =>
+  returnNameArrayIfNotEmpty(
+    state.dashboardElement && state.dashboardElement.commoditiesPanel.activeItems
+  );
+const getSourcesPanel = state =>
+  returnNameArrayIfNotEmpty(
+    state.dashboardElement && state.dashboardElement.commoditiesPanel.activeItems
+  );
+
+const getVariableColumnName = createSelector(
+  [getMeta],
+  meta => (meta.info.node_type === 'COUNTRY' ? 'DESTINATION' : meta.info.node_type)
+);
+
+const hasNContIndicator = createSelector(
+  [getMeta],
+  meta => meta.info.filter.ncont_attribute
+);
+const hasMultipleYears = createSelector(
+  [getMeta],
+  meta => meta.info.years.end_year
+);
+const hasVariableColumn = createSelector(
+  [getChartType, getVariableColumnName, hasMultipleYears, hasNContIndicator, getSourcesPanel],
+  (chartType, variableColumnName, _hasMultipleYears, _hasNContIndicator, sourcesPanel) =>
+    !(!sourcesPanel && _hasNContIndicator && _hasMultipleYears) &&
+    !(!sourcesPanel && chartType === 'bar' && _hasMultipleYears) &&
+    !(chartType === 'pie') &&
+    variableColumnName
+);
+
+const getIndicatorColumnName = (meta, type) => {
+  const unit = type === 'cont' && (meta.xAxis.suffix || meta.yAxis.suffix);
+  return `${meta.info.filter[`${type}_attribute`].toUpperCase()}${unit ? ` (${unit})` : ''}`;
+};
+
+export const getTableHeaders = createSelector(
+  [getMeta, hasVariableColumn, getVariableColumnName, hasNContIndicator],
+  (meta, _hasVariableColumn, variableColumnName, _hasNContIndicator) => {
+    if (!meta) return null;
+    const headers = ['COMMODITY', 'COUNTRY', 'YEAR'];
+    if (_hasVariableColumn) headers.push(variableColumnName);
+    headers.push(getIndicatorColumnName(meta, 'cont'));
+    if (_hasNContIndicator) {
+      headers.push(getIndicatorColumnName(meta, 'ncont'));
+    }
+    return headers;
+  }
+);
+
+const getYear = (d, meta, chartType) => {
+  switch (chartType) {
+    case CHART_TYPES.horizontalBar:
+    case CHART_TYPES.horizontalStackedBar:
+    case CHART_TYPES.pie:
+      return meta.info.years.start_year;
+    default:
+      return d.x || meta.info.years.start_year;
+  }
+};
+
+const getContValue = (d, c) => d[c] && format(',.2s')(d[c]);
+const getNotContValue = (d, c, meta, chartType, _hasMultipleYears) => {
+  if (
+    chartType === CHART_TYPES.horizontalBar ||
+    chartType === CHART_TYPES.horizontalStackedBar ||
+    _hasMultipleYears
+  ) {
+    return meta[c].label && meta[c].label.toUpperCase();
```

Comment:
what does `c` stands for? I prefer using less abbreviations when possible

Reply:
d is a unit of data, can you find a better name?